src/model/level/internal.rs

```diff
@@ -0,0 +1,332 @@
+    pub main_song: MainSong,
+
+    #[serde(rename = "13")]
+    pub gd_version: u8,
+
+    #[serde(rename = "14")]
+    pub likes: i32,
+
+    #[serde(rename = "15", with = "level_length")]
+    pub length: LevelLength,
+
+    #[serde(rename = "18")]
+    pub stars: u8,
+
+    #[serde(rename = "19")]
+    pub featured: Featured,
+
+    #[serde(rename = "30", with = "crate::util::default_to_none")]
+    pub copy_of: Option<u64>,
+
+    #[serde(rename = "31")]
+    pub index_31: Option<&'a str>,
+
+    #[serde(rename = "35", with = "crate::util::default_to_none")]
+    pub custom_song: Option<u64>,
+
+    #[serde(rename = "37")]
+    pub coin_amount: u8,
+
+    #[serde(rename = "38")]
+    pub coins_verified: bool,
+
+    #[serde(rename = "39", with = "crate::util::default_to_none")]
+    pub stars_requested: Option<u8>,
+
+    #[serde(rename = "40")]
+    pub index_40: Option<&'a str>,
+
+    #[serde(rename = "42")]
+    pub is_epic: bool,
+
+    #[serde(rename = "43")]
+    pub index_43: &'a str,
+
+    #[serde(rename = "45", with = "crate::util::default_to_none")]
+    pub object_amount: Option<u32>,
+
+    #[serde(rename = "46")]
+    pub index_46: Option<&'a str>,
+
+    #[serde(rename = "47")]
+    pub index_47: Option<&'a str>,
+
+    #[serde(rename = "4", default, skip_serializing_if = "Option::is_none")]
+    pub level_data: Option<&'a str>,
+
+    #[serde(rename = "27", default, skip_serializing_if = "Option::is_none")]
+    pub password: Option<Internal<Password>>,
+
+    #[serde(rename = "28", default, skip_serializing_if = "Option::is_none")]
+    pub time_since_upload: Option<&'a str>,
+
+    #[serde(rename = "29", default, skip_serializing_if = "Option::is_none")]
+    pub time_since_update: Option<&'a str>,
+
+    #[serde(rename = "36", default, skip_serializing_if = "Option::is_none")]
+    pub index_36: Option<&'a str>,
+}
+
+impl<'a> HasRobtopFormat<'a> for Level<'a, Option<u64>, u64> {
+    type Internal = InternalLevel<'a>;
+
+    const DELIMITER: &'static str = ":";
+    const MAP_LIKE: bool = true;
+
+    fn as_internal(&'a self) -> Self::Internal {
+        InternalLevel {
+            level_id: self.level_id,
+            name: self.name.borrow(),
+            description: self.description.as_ref().map(|thunk| {
+                Internal(match thunk {
+                    Thunk::Unprocessed(unproc) => Thunk::Unprocessed(unproc),
+                    Thunk::Processed(Base64Decoded(moo)) => Thunk::Processed(Base64Decoded(Cow::Borrowed(moo.as_ref()))),
+                })
+            }),
+            version: self.version,
+            creator: self.creator,
+            is_auto: self.difficulty == LevelRating::Auto,
+            has_difficulty_rating: self.difficulty != LevelRating::NotAvailable,
+            rating: self.difficulty.into_response_value(),
+            is_demon: self.difficulty.is_demon(),
+            downloads: self.downloads,
+            main_song: self.main_song.unwrap_or(MAIN_SONGS[0]),
```

Comment:
Consider changing MAIN_SONGS[0] into an embedded constant so in case the array changes or for any other reason the index fails we catch it at compile time.
The solution suggested would be to define a const FIRST_SONG inside this function referring to MAIN_SONGS[0] and using that

Reply:
Mh, serialization here is only based on the song ID, so the only purpose of using MAIN_SONGS[0] here is to have the field serialize to zero.
Thinking about it now, it's probably better to just store only the main song ID in InternalLevel

Reply:
Yeee that's definitely a better way because then MainSong's (De)Serialize implementation does not look implementation details (rn its implemented via song id)